src/Conversion/ONNXToKrnl/Tensor/Reshape.cpp

```diff
@@ -31,71 +61,73 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
+      // other dimensions. But we don't have enough information about the other
+      // dimensions at this point. So, we need to scan the shape first to
+      // calculate reduction of all of the dimensions. If the reduction is
+      // negative, then the shape array contains a negative dimension.
+      // Otherwise, the reduction is the same as the one computed from the input
+      // tensor.
       Value tensorSizeFromShape = emitConstantOp(rewriter, loc,
           rewriter.getIntegerType(64), getMemRefEltSizeInBytes(memRefType));
       SmallVector<Value, 4> DimInfo;
       for (int i = 0; i < memRefShape.size(); ++i) {
-        Value index = emitConstantOp(rewriter, loc, rewriter.getIndexType(), i);
-        // Load index from array of indices.
-        Value loadedVal = rewriter.create<KrnlLoadOp>(loc, operands[1], index);
-        // If a dimension is zero, the actual dimension value is taken from the
-        // input tensor.
-        //
-        // If a dimension is negative, it is computed from the other dimensions.
+        // If a dimension is N (N != 0 && N != -1), use it.
+        // If a dimension is - 1, it will be computed from the other dimensions.
         // But we don't have enough information about the other dimensions at
         // this point. So, we let it as it is (-1), and compute it later.
-        if (i < inputShape.size()) {
-          Value dimVal;
-          auto loadedValType = loadedVal.getType().cast<IntegerType>();
-          if (inputShape[i] < 0) {
-            Value dim = rewriter.create<DimOp>(loc, data, i);
-            dimVal = rewriter.create<IndexCastOp>(loc, dim, loadedValType);
+        Value loadedVal;
+        if (!shapeAttrValues.empty()) {
+          loadedVal = emitConstantOp(
+              rewriter, loc, rewriter.getIntegerType(64), shapeAttrValues[i]);
+        } else {
+          Value index =
+              emitConstantOp(rewriter, loc, rewriter.getIndexType(), i);
+          loadedVal = rewriter.create<KrnlLoadOp>(loc, shape, index);
+        }
+
+        // A dimension cannot be 0 if its position is out-of-bound, e.g. the
+        // output rank is greater than the input rank.
+        if (i < dataShape.size()) {
+          // If a dimension is 0, the actual dimension value is taken from the
+          // input tensor.
+          if (!shapeAttrValues.empty()) {
+            if (shapeAttrValues[i] == 0)
+              loadedVal = getDimOrConstant(rewriter, loc, data, i);
           } else {
-            dimVal =
-                emitConstantOp(rewriter, loc, loadedValType, inputShape[i]);
+            auto dimVal = getDimOrConstant(rewriter, loc, data, i);
+            auto zero =
+                emitConstantOp(rewriter, loc, rewriter.getIntegerType(64), 0);
+            auto isZero = rewriter.create<CmpIOp>(
+                loc, CmpIPredicate::eq, loadedVal, zero);
+            loadedVal =
+                rewriter.create<SelectOp>(loc, isZero, dimVal, loadedVal);
```

Comment:
The loadedValue are assigned multiple times and its semantic changes in the process. Use different name, such as loadedValue then outputDimValue? 
The loadedValue is compared with 0, how about -1?  Will it make the tensorSizeFromShape become negative?

Reply:
@chentong319 thanks for your comments!
I reorganized the code to make it clear. 

> how about -1? Will it make the tensorSizeFromShape become negative?

Yes, it becomes negative. We need to make it positive later.

> You may test the code with promoteConstantToAttribute disabled.
it is already in the onnx_lowering.mlir test.